### PR TITLE
etcd-tester: fix tester

### PR DESCRIPTION
**Pull request by gyuho**
Fix coreos#5281.
Fix coreos#5317.

### CODE REVIEW

**Comment by gyuho** [comment]
@xiang90 I think this fixes the issue. Been running over 100 rounds without failure:

- restart stresser only when revision check succeeds
- in case compact request times out, do retries

It sometimes happens when we inject network latency. The request went through and got processed but timed out, as below:

```
2016-05-06 23:10:09.250670 I | etcd-tester: [round#16] compacting storage at 33036779 (current revision 33046779)
2016-05-06 23:10:39.250898 I | etcd-tester: [compactKV #0] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-06 23:11:09.251356 I | etcd-tester: [compactKV #1] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-06 23:11:39.251632 I | etcd-tester: [compactKV #2] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-06 23:12:02.465257 I | etcd-tester: 10.240.0.17:2379 is already compacted with 33036779 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-06 23:12:06.043466 I | etcd-tester: 10.240.0.18:2379 is already compacted with 33036779 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-06 23:12:06.152338 I | etcd-tester: 10.240.0.19:2379 is already compacted with 33036779 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)

2016-05-07 00:33:34.771060 I | etcd-tester: [compactKV #0] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-07 00:34:04.771503 I | etcd-tester: [compactKV #1] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-07 00:34:34.772002 I | etcd-tester: [compactKV #2] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-07 00:35:04.772377 I | etcd-tester: [compactKV #3] kvc.Compact 10.240.0.17:2379 error rpc error: code = 4 desc = context deadline exceeded
2016-05-07 00:35:14.679767 I | etcd-tester: 10.240.0.17:2379 is already compacted with 53540230 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-07 00:35:16.711301 I | etcd-tester: 10.240.0.18:2379 is already compacted with 53540230 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
2016-05-07 00:35:21.600652 I | etcd-tester: 10.240.0.19:2379 is already compacted with 53540230 (rpc error: code = 11 desc = etcdserver: mvcc: required revision has been compacted)
```

- other minor fixes

Thanks.

**Comment by gyuho** [comment]
Seen other failures from recovery request timeout. Investigating...

**Comment by gyuho** [comment]
@xiang90 I've been running this patch in a separate testing cluster over 1,000 rounds without failure. This should fix those issues.

PTAL.

File: tools/functional-tester/etcd-tester/cluster.go [hunk 1]
```diff
@@ -175,6 +175,7 @@ func (c *cluster) WaitHealth() error {
 		if err == nil {
 			return nil
 		}
+		plog.Printf("[WaitHealth #%d] setHealthKey error %q", i, err)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: plog.Warnf

File: tools/functional-tester/etcd-tester/cluster.go [hunk 2]
```diff
@@ -261,17 +262,17 @@ func (c *cluster) Status() ClusterStatus {
 // setHealthKey sets health key on all given urls.
 func setHealthKey(us []string) error {
 	for _, u := range us {
-		conn, err := grpc.Dial(u, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
+		conn, err := grpc.Dial(u, grpc.WithInsecure(), grpc.WithTimeout(15*time.Second))
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: why 15 seconds instead of 5?

File: tools/functional-tester/etcd-tester/cluster.go [hunk 3]
```diff
@@ -261,17 +262,17 @@ func (c *cluster) Status() ClusterStatus {
...
 		if err != nil {
 			return fmt.Errorf("%v (%s)", err, u)
 		}
-		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/cluster.go]: why 10 seconds?

File: tools/functional-tester/etcd-tester/tester.go [hunk 4]
```diff
@@ -41,84 +41,87 @@ func (tt *tester) runLoop() {
 		tt.status.setRound(round)
 		tt.status.setCase(-1) // -1 so that logPrefix doesn't print out 'case'
 		roundTotalCounter.Inc()
+		pfx0 := tt.logPrefix()
 
 		var failed bool
 		for j, f := range tt.failures {
 			caseTotalCounter.WithLabelValues(f.Desc()).Inc()
 			tt.status.setCase(j)
+			pfx1 := tt.logPrefix()
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/tester.go]: what the difference between pfx1 and fpx0? i feel they are bad names. we need to be more descriptive.

**Comment by gyuho** [comment]
@xiang90 PTAL.

Reverted timeout changes (turns out it wasn't needed), and made the variable names more descriptive.

Thanks.

**Comment by gyuho** [comment]
Kept seeing context deadline exceeded... I will bisect old changes to find what I did wrong before this PR.

**Comment by gyuho** [comment]
Closing this in favor of https://github.com/coreos/etcd/pull/5421.

**Pull request closed without merging**